README.md
# Mountall enhancements for ZFS on Linux

The home for this git repository is:

* https://github.com/zfsonlinux/mountall

Mountall packages in the ZoL PPA at https://launchpad.net/~zfs-native are built
from this repository using the git-buildpackage tool.


## Casual Build Instructions

Just do quick builds like this:
```
$ apt-get source --build mountall
```

## Developer Build Instructions

1. Clone this repository:
```
$ git clone git://github.com/zfsonlinux/mountall.git
$ cd mountall
```

1. List the current releases by branch name:
```
$ git branch --list 'master/*'
```

1. Or list previous releases by tag name:
```
$ git tag --list 'master/*
```

1. Checkout the branch name or tag name that you want to build.  For example,
the latest code for Ubuntu 12.04 Precise Pangolin is:
```
$ git checkout master/ubuntu/precise
```

1. Now compile it:
```
$ git-buildpackage -uc -us
```

1. And clean the working tree afterwards by doing this:
```
$ git clean -df
$ git reset --hard
```

## Release Build Instructions

```
$ git-buildpackage --git-tag -S -sa
$ dput ppa:zfs-native/staging ../mountall_*_source.changes
```

## Upstream Repositories

Each of the `upstream` branches in this repository is an unmodified copy of an
official distribution repository, like those hosted by the Debian or Ubuntu
projects.  The `git-buildpackage` framework combines an `upstream` branch with
the corresponding `patch-queue` branch to create a `master` branch according to
the `debian/gbp.conf` file.

By default, git will pull the `upstream` branches from the clone origin at
Github and not the actual upstream respositories.  Look at the `git-config.txt`
file in the `readme` branch for an example of how to configure remotes that
pull from the official distribution repositories.

Some of the remotes for this repository are bzr repositories at Launchpad.
Pulling through bzr requires this bridge helper:

https://github.com/felipec/git/blob/fc/remote/bzr/contrib/remote-helpers/git-remote-bzr

Just download that file directly into the `/usr/local/bin` directory and make
it executable.  Also run `bzr launchpad-login` to ensure that you have a
working bazaar installation.

The `upstream` branches for Ubuntu may be disconnected or incomplete because
Ubuntu does not always use source control for system updates.  In such cases,
the `upstream` branches will be supplemented by `git-import-orig`.


debian/changelog
mountall (2.54ubuntu1-zfs1) vivid; urgency=low

  * Import mountall 2.53-zfs1 changes for ZoL.
  * Update gbp.conf for Ubuntu 15.04 Vivid Vervet.
  * git-import-dsc mountall_2.54ubuntu1.dsc

 -- Darik Horn <dajhorn@vanadac.com>  Sat, 21 Feb 2015 16:57:47 -0500

mountall (2.54ubuntu1) vivid; urgency=medium

  * SECURITY UPDATE: insecure mount permissions (LP: #1390183)
    - The mount utility now honours process umask when mounting certain
      filesystems, resulting in them being potentially mounted with
      inappropriate permissions.
    - src/mountall.c: don't specifically set umask when running as a
      daemon, inherit the umask Upstart sets instead.
    - CVE-2014-1421

 -- Marc Deslauriers <marc.deslauriers@ubuntu.com>  Thu, 13 Nov 2014 13:11:38 -0500

mountall (2.54build1) utopic; urgency=high

  * No change rebuild against libplymouth4.

 -- Dimitri John Ledkov <xnox@ubuntu.com>  Mon, 14 Jul 2014 10:15:29 +0100

mountall (2.54) unstable; urgency=medium

  * conf/mounted-proc.conf: /proc/slabinfo may not exist, in which case we
    should avoid trying to chmod it and generating spurious errors in the
    log.  LP: #1319724.

 -- Steve Langasek <vorlon@debian.org>  Thu, 15 May 2014 07:54:40 -0700

mountall (2.53) unstable; urgency=medium

  [ James Hunt ]
  * man/mountall.8:
    - Added SEE ALSO section.
    - Updated date and copyright.
    - Remove comments about mountall being a temporary tool
      since this is not true (it's been "temporary" since 2009) and it
      undermines its current importance.
    - Reworked DESCRIPTION and made mention of /lib/init/fstab.

  [ Steve Langasek ]
  * conf/mounted-run.conf: the dynamic motd location used by PAM has changed
    to /run/motd.dynamic; update the job to match.

 -- Steve Langasek <vorlon@debian.org>  Wed, 19 Feb 2014 16:17:14 -0800

mountall (2.52) unstable; urgency=low

  * Don't emit extra 'mounting' events for mounts already in progress; this
    will cause double triggering of some jobs for remote filesystems, and
    can also cause us to miss 'mounted' events.  LP: #1048017.
  * Fix mountall upstart job to not start a subshell for reading
    /proc/cmdline, since this causes upstart to lose track of the daemon
    process and leaves mountall-net unable to signal it to retry network
    mounts.  LP: #1235013.

 -- Steve Langasek <vorlon@debian.org>  Wed, 09 Oct 2013 04:12:51 +0000

mountall (2.51) unstable; urgency=low

  * Fix tagging of filesystems to not have local/remote inheritance
    overridden; otherwise we will mis-tag various mounts and deadlock the
    boot.  Also fixes an inconsistency with the inheritance of
    'bootwait'/'nobootwait' flags depending on the order of mounts in
    /etc/fstab: we now always treat the 'nobootwait' flag as applying to
    submounts. LP: #1223745, LP: #1153672.

 -- Steve Langasek <vorlon@debian.org>  Fri, 13 Sep 2013 22:23:55 -0700

mountall (2.50) unstable; urgency=low

  * Add support for pstore.  Closes: #722212.
  * Allow multiple fstab entries with the same mountpoint: we still mask
    any entries that came from a different source (i.e., /etc/fstab will
    mask /lib/init/fstab), but if there are multiple entries in /etc/fstab,
    treat them all as separate mounts instead of letting the last one win.
    LP: #503003.
  * Don't set 'console output' anymore; now that we have upstart logging by
    default, logs will be captured and (assuming /var/log becomes writable
    at some point) mountall is debuggable without having to spew text to the
    console.
  * Support reading --verbose/--debug options from /proc/cmdline, so that
    we can give users a better way to debug mountall without having to
    edit the upstart job.
  * Since everything is in a single event loop, a request from plymouth to
    skip a mount may come in after the mount has already happened.  So don't
    assert on a bogus request, just treat it as a no-op and continue.
    LP: #731800.

 -- Steve Langasek <vorlon@debian.org>  Tue, 10 Sep 2013 04:01:19 +0000

mountall (2.49) unstable; urgency=low

  [ James Hunt ]
  * apport/mountall.py: add a mountall apport hook.

  [ Steve Langasek ]
  * Ignore parse errors in /etc/default/rcS or /etc/default/locale; these
    problems shouldn't be allowed to prevent the system from booting.
    LP: #1192514.

 -- Steve Langasek <vorlon@debian.org>  Tue, 02 Jul 2013 15:51:12 -0700

mountall (2.48) unstable; urgency=low

  [ Steve Langasek ]
  * Make sure we don't show the error message claiming a device is not
    ready when we're actually already handling it but are just waiting for
    related upstart jobs that are blocking the mount to finish.
    LP: #1091792.

  [ Stéphane Graber ]
  * Mount a tmpfs on /sys/fs/cgroup if it exists.

 -- Steve Langasek <vorlon@debian.org>  Sat, 09 Mar 2013 00:36:22 +0000

mountall (2.47) unstable; urgency=low

  [ Dave Chiluk ]
  * Adjust parsing of options so mountall doesn't strip options that are
    substrings of these strings (showthrough, optional, bootwait, nobootwait
    or timeout). This fixes the issue where timeo was getting stripped from
    nfs mounts. LP: #1041377.

  [ Steve Langasek ]
  * Ensure callbacks are called directly when running with --no-events,
    otherwise the "event" handling of the non-events never finishes and
    mountall hangs.  LP: #1099349.

 -- Steve Langasek <vorlon@debian.org>  Mon, 14 Jan 2013 16:37:36 -0800

mountall (2.46) unstable; urgency=low

  * Fix a further remaining issue from 2.41: by fixing the missing 'mounted'
    events in 2.43, we have again introduced a case where mountall may block
    waiting for 'mounted MOUNTPOINT=/' to return before handling 'mounted
    MOUNTPOINT=/run' and emitting virtual-filesystems, thus causing
    dependency loops and long boot timeouts.  LP: #1078926.

 -- Steve Langasek <vorlon@debian.org>  Mon, 03 Dec 2012 23:44:16 +0000

mountall (2.45) unstable; urgency=low

  [ Serge Hallyn ]

config.h
/* config.h.  Generated from config.h.in by configure.  */
/* config.h.in.  Generated from configure.ac by autoheader.  */

/* Define to one of `_getb67', `GETB67', `getb67' for Cray-2 and Cray-YMP
   systems. This function is required for `alloca.c' support on those systems.
   */
/* #undef CRAY_STACKSEG_END */

/* Define to 1 if using `alloca.c'. */
/* #undef C_ALLOCA */

/* Define to 1 if translation of program messages to the user's native
   language is requested. */
#define ENABLE_NLS 1

/* Define to 1 if you have `alloca', as a function or macro. */
#define HAVE_ALLOCA 1

/* Define to 1 if you have <alloca.h> and it should be used (not on Ultrix).
   */
#define HAVE_ALLOCA_H 1

/* Define to 1 if you have the `argz_count' function. */
#define HAVE_ARGZ_COUNT 1

/* Define to 1 if you have the <argz.h> header file. */
#define HAVE_ARGZ_H 1

/* Define to 1 if you have the `argz_next' function. */
#define HAVE_ARGZ_NEXT 1

/* Define to 1 if you have the `argz_stringify' function. */
#define HAVE_ARGZ_STRINGIFY 1

/* Define to 1 if you have the `asprintf' function. */
#define HAVE_ASPRINTF 1

/* Define to 1 if the compiler understands __builtin_expect. */
#define HAVE_BUILTIN_EXPECT 1

/* Define to 1 if you have the MacOS X function CFLocaleCopyCurrent in the
   CoreFoundation framework. */
/* #undef HAVE_CFLOCALECOPYCURRENT */

/* Define to 1 if you have the MacOS X function CFPreferencesCopyAppValue in
   the CoreFoundation framework. */
/* #undef HAVE_CFPREFERENCESCOPYAPPVALUE */

/* Define if the GNU dcgettext() function is already present or preinstalled.
   */
#define HAVE_DCGETTEXT 1

/* Define to 1 if you have the declaration of `feof_unlocked', and to 0 if you
   don't. */
#define HAVE_DECL_FEOF_UNLOCKED 1

/* Define to 1 if you have the declaration of `fgets_unlocked', and to 0 if
   you don't. */
#define HAVE_DECL_FGETS_UNLOCKED 1

/* Define to 1 if you have the declaration of `getc_unlocked', and to 0 if you
   don't. */
#define HAVE_DECL_GETC_UNLOCKED 1

/* Define to 1 if you have the declaration of `_snprintf', and to 0 if you
   don't. */
#define HAVE_DECL__SNPRINTF 0

/* Define to 1 if you have the declaration of `_snwprintf', and to 0 if you
   don't. */
#define HAVE_DECL__SNWPRINTF 0

/* Define to 1 if you have the <dlfcn.h> header file. */
#define HAVE_DLFCN_H 1

/* Define to 1 if you have the `fwprintf' function. */
#define HAVE_FWPRINTF 1

/* Define to 1 if you have the `getcwd' function. */
#define HAVE_GETCWD 1

/* Define to 1 if you have the `getegid' function. */
#define HAVE_GETEGID 1

/* Define to 1 if you have the `geteuid' function. */
#define HAVE_GETEUID 1

/* Define to 1 if you have the `getgid' function. */
#define HAVE_GETGID 1

/* Define to 1 if you have the `getpagesize' function. */
#define HAVE_GETPAGESIZE 1

/* Define if the GNU gettext() function is already present or preinstalled. */
#define HAVE_GETTEXT 1

/* Define to 1 if you have the `getuid' function. */
#define HAVE_GETUID 1

/* Define if you have the iconv() function. */
#define HAVE_ICONV 1

/* Define if you have the 'intmax_t' type in <stdint.h> or <inttypes.h>. */
#define HAVE_INTMAX_T 1

/* Define to 1 if you have the <inttypes.h> header file. */
#define HAVE_INTTYPES_H 1

/* Define if <inttypes.h> exists, doesn't clash with <sys/types.h>, and
   declares uintmax_t. */
#define HAVE_INTTYPES_H_WITH_UINTMAX 1

/* Define if you have <langinfo.h> and nl_langinfo(CODESET). */
#define HAVE_LANGINFO_CODESET 1

/* Define if your <locale.h> file defines LC_MESSAGES. */
#define HAVE_LC_MESSAGES 1

/* Define to 1 if you have the <limits.h> header file. */
#define HAVE_LIMITS_H 1

/* Define if you have the 'long double' type. */
#define HAVE_LONG_DOUBLE 1

/* Define to 1 if the system has the type `long long int'. */
#define HAVE_LONG_LONG_INT 1

/* Define to 1 if you have the <memory.h> header file. */
#define HAVE_MEMORY_H 1

/* Define to 1 if you have the `mempcpy' function. */
#define HAVE_MEMPCPY 1

/* Define to 1 if you have a working `mmap' system call. */
#define HAVE_MMAP 1

/* Define to 1 if you have the `munmap' function. */
#define HAVE_MUNMAP 1

/* Define if you have <langinfo.h> and it defines the NL_LOCALE_NAME macro if
   _GNU_SOURCE is defined. */
#define HAVE_NL_LOCALE_NAME 1

/* Define if your printf() function supports format strings with positions. */
#define HAVE_POSIX_PRINTF 1

/* Define if the <pthread.h> defines PTHREAD_MUTEX_RECURSIVE. */
#define HAVE_PTHREAD_MUTEX_RECURSIVE 1

/* Define if the POSIX multithreading library has read/write locks. */
#define HAVE_PTHREAD_RWLOCK 1

/* Define to 1 if you have the `putenv' function. */
#define HAVE_PUTENV 1

/* Define to 1 if you have the `setenv' function. */
#define HAVE_SETENV 1

/* Define to 1 if you have the `setlocale' function. */
#define HAVE_SETLOCALE 1

/* Define to 1 if you have the `snprintf' function. */
#define HAVE_SNPRINTF 1

/* Define to 1 if you have the <stddef.h> header file. */
#define HAVE_STDDEF_H 1

/* Define to 1 if you have the <stdint.h> header file. */
#define HAVE_STDINT_H 1

/* Define if <stdint.h> exists, doesn't clash with <sys/types.h>, and declares
   uintmax_t. */
#define HAVE_STDINT_H_WITH_UINTMAX 1

/* Define to 1 if you have the <stdlib.h> header file. */
#define HAVE_STDLIB_H 1

/* Define to 1 if you have the `stpcpy' function. */
#define HAVE_STPCPY 1

/* Define to 1 if you have the `strcasecmp' function. */
#define HAVE_STRCASECMP 1

/* Define to 1 if you have the `strdup' function. */
#define HAVE_STRDUP 1

/* Define to 1 if you have the <strings.h> header file. */
#define HAVE_STRINGS_H 1

/* Define to 1 if you have the <string.h> header file. */
#define HAVE_STRING_H 1

/* Define to 1 if you have the `strtoul' function. */
#define HAVE_STRTOUL 1

/* Define to 1 if you have the <sys/param.h> header file. */
#define HAVE_SYS_PARAM_H 1

/* Define to 1 if you have the <sys/stat.h> header file. */
#define HAVE_SYS_STAT_H 1

/* Define to 1 if you have the <sys/types.h> header file. */
#define HAVE_SYS_TYPES_H 1

/* Define to 1 if you have the `tsearch' function. */
#define HAVE_TSEARCH 1

/* Define if you have the 'uintmax_t' type in <stdint.h> or <inttypes.h>. */
#define HAVE_UINTMAX_T 1

/* Define to 1 if you have the <unistd.h> header file. */
#define HAVE_UNISTD_H 1

/* Define if you have the 'unsigned long long' type. */
#define HAVE_UNSIGNED_LONG_LONG 1

/* Define to 1 if the system has the type `unsigned long long int'. */
#define HAVE_UNSIGNED_LONG_LONG_INT 1

/* Define to 1 if you have the <valgrind/valgrind.h> header file. */
#define HAVE_VALGRIND_VALGRIND_H 1

/* Define to 1 or 0, depending whether the compiler supports simple visibility
   declarations. */
#define HAVE_VISIBILITY 1

/* Define if you have the 'wchar_t' type. */
#define HAVE_WCHAR_T 1

/* Define to 1 if you have the `wcslen' function. */
#define HAVE_WCSLEN 1

/* Define if you have the 'wint_t' type. */
#define HAVE_WINT_T 1

/* Define to 1 if you have the `__fsetlocking' function. */
#define HAVE___FSETLOCKING 1

/* Define as const if the declaration of iconv() needs const. */
#define ICONV_CONST 

/* Define if integer division by zero raises signal SIGFPE. */
#define INTDIV0_RAISES_SIGFPE 1

/* Define to the sub-directory in which libtool stores uninstalled libraries.
   */
#define LT_OBJDIR ".libs/"

/* Define to 1 if your C compiler doesn't accept -c and -o together. */
/* #undef NO_MINUS_C_MINUS_O */

/* Name of package */
#define PACKAGE "mountall"

/* Define to the address where bug reports for this package should be sent. */
#define PACKAGE_BUGREPORT "ubuntu-devel@lists.ubuntu.com"

/* Define to the copyright message of this package. */
#define PACKAGE_COPYRIGHT "Copyright (C) 2009 Canonical Ltd."

/* Define to the full name of this package. */
#define PACKAGE_NAME "mountall"

/* Define to the full name and version of this package. */
#define PACKAGE_STRING "mountall 1.0"

/* Define to the one symbol short name of this package. */
#define PACKAGE_TARNAME "mountall"

/* Define to the home page for this package. */
#define PACKAGE_URL ""

/* Define to the version of this package. */
#define PACKAGE_VERSION "1.0"

/* Define if <inttypes.h> exists and defines unusable PRI* macros. */
/* #undef PRI_MACROS_BROKEN */

/* Define if the pthread_in_use() detection is hard. */
/* #undef PTHREAD_IN_USE_DETECTION_HARD */

/* Define as the maximum value of type 'size_t', if the system doesn't define
   it. */
/* #undef SIZE_MAX */

/* If using the C implementation of alloca, define if you know the
   direction of stack growth for your system; otherwise it will be
   automatically deduced at runtime.
	STACK_DIRECTION > 0 => grows toward higher addresses
	STACK_DIRECTION < 0 => grows toward lower addresses
	STACK_DIRECTION = 0 => direction of growth unknown */
/* #undef STACK_DIRECTION */

/* Define to 1 if you have the ANSI C header files. */
#define STDC_HEADERS 1

/* Define if the POSIX multithreading library can be used. */
#define USE_POSIX_THREADS 1

/* Define if references to the POSIX multithreading library should be made
   weak. */
#define USE_POSIX_THREADS_WEAK 1

/* Define if the GNU Pth multithreading library can be used. */
/* #undef USE_PTH_THREADS */

/* Define if references to the GNU Pth multithreading library should be made
   weak. */
/* #undef USE_PTH_THREADS_WEAK */

/* Define if the old Solaris multithreading library can be used. */
/* #undef USE_SOLARIS_THREADS */

/* Define if references to the old Solaris multithreading library should be
   made weak. */
/* #undef USE_SOLARIS_THREADS_WEAK */

/* Enable extensions on AIX 3, Interix.  */
#ifndef _ALL_SOURCE
# define _ALL_SOURCE 1
#endif
/* Enable GNU extensions on systems that have them.  */
#ifndef _GNU_SOURCE
# define _GNU_SOURCE 1
#endif
/* Enable threading extensions on Solaris.  */
#ifndef _POSIX_PTHREAD_SEMANTICS
# define _POSIX_PTHREAD_SEMANTICS 1
#endif
/* Enable extensions on HP NonStop.  */
#ifndef _TANDEM_SOURCE
# define _TANDEM_SOURCE 1
#endif
/* Enable general extensions on Solaris.  */
#ifndef __EXTENSIONS__
# define __EXTENSIONS__ 1
#endif


/* Define if the Win32 multithreading API can be used. */
/* #undef USE_WIN32_THREADS */

/* Version number of package */
#define VERSION "1.0"

/* Number of bits in a file offset, on hosts where this is settable. */
#define _FILE_OFFSET_BITS 64

/* Define for large files, on AIX-style hosts. */
/* #undef _LARGE_FILES */

/* Define to 1 if on MINIX. */
/* #undef _MINIX */

/* Define to 2 if the system does not provide POSIX.1 features except with
   this defined. */
/* #undef _POSIX_1_SOURCE */

/* Define to 1 if you need to in order for `stat' and other things to work. */
/* #undef _POSIX_SOURCE */

/* Define to empty if `__thread' is not supported. */
#define __thread /**/

/* Define to `__inline__' or `__inline' if that's what the C compiler
   calls it, or to nothing if 'inline' is not supported under any name.  */
#ifndef __cplusplus
/* #undef inline */
#endif

/* Define as the type of the result of subtracting two pointers, if the system
   doesn't define it. */
/* #undef ptrdiff_t */

/* Define to `unsigned int' if <sys/types.h> does not define. */
/* #undef size_t */

/* Define to unsigned long or unsigned long long if <stdint.h> and
   <inttypes.h> don't define. */
/* #undef uintmax_t */


#define __libc_lock_t                   gl_lock_t
#define __libc_lock_define              gl_lock_define
#define __libc_lock_define_initialized  gl_lock_define_initialized
#define __libc_lock_init                gl_lock_init
#define __libc_lock_lock                gl_lock_lock
#define __libc_lock_unlock              gl_lock_unlock
#define __libc_lock_recursive_t                   gl_recursive_lock_t
#define __libc_lock_define_recursive              gl_recursive_lock_define
#define __libc_lock_define_initialized_recursive  gl_recursive_lock_define_initialized
#define __libc_lock_init_recursive                gl_recursive_lock_init
#define __libc_lock_lock_recursive                gl_recursive_lock_lock
#define __libc_lock_unlock_recursive              gl_recursive_lock_unlock
#define glthread_in_use  libintl_thread_in_use
#define glthread_lock_init     libintl_lock_init
#define glthread_lock_lock     libintl_lock_lock
#define glthread_lock_unlock   libintl_lock_unlock
#define glthread_lock_destroy  libintl_lock_destroy
#define glthread_rwlock_init     libintl_rwlock_init
#define glthread_rwlock_rdlock   libintl_rwlock_rdlock
#define glthread_rwlock_wrlock   libintl_rwlock_wrlock
#define glthread_rwlock_unlock   libintl_rwlock_unlock
#define glthread_rwlock_destroy  libintl_rwlock_destroy
#define glthread_recursive_lock_init     libintl_recursive_lock_init
#define glthread_recursive_lock_lock     libintl_recursive_lock_lock
#define glthread_recursive_lock_unlock   libintl_recursive_lock_unlock
#define glthread_recursive_lock_destroy  libintl_recursive_lock_destroy
#define glthread_once                 libintl_once
#define glthread_once_call            libintl_once_call
#define glthread_once_singlethreaded  libintl_once_singlethreaded



src/control.h
/* mountall 
 *
 * Copyright © 2010 Canonical Ltd.
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License version 2, as
 * published by the Free Software Foundation.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License along
 * with this program; if not, write to the Free Software Foundation, Inc.,
 * 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
 */

#include "dbus/mountall.h"

extern DBusServer *control_server;
extern const NihDBusInterface *control_interfaces[];

int control_server_open (void);
int control_stop_timer  (void *data, NihDBusMessage *message, 
						const char *mountpoint)
	__attribute__ ((warn_unused_result));
int control_restart_timer (void *data, NihDBusMessage *message,
	       					const char *mountpoint)
	__attribute__ ((warn_unused_result));
int control_change_mount_device (void *data, NihDBusMessage *message, 
				const char *devname, const char *path)
	__attribute__ ((warn_unused_result));
int control_get_version (void *data, NihDBusMessage *message, 
						char **value)
	__attribute__ ((warn_unused_result));


src/mountall.h
/* mountall 
 *
 * Copyright © 2010 Canonical Ltd.
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License version 2, as
 * published by the Free Software Foundation.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License along
 * with this program; if not, write to the Free Software Foundation, Inc.,
 * 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
 */

int stop_dev_timer (const char * mountpoint)
	__attribute__ ((warn_unused_result));
int restart_dev_timer (const char * mountpoint)
	__attribute__ ((warn_unused_result));
int change_mount_device (const char * devname, const char * path)
	__attribute__ ((warn_unused_result));


util/mntctl.c
/* mountall
 *
 * Copyright © 2010 Canonical Ltd.
 *
 * This file is based on initctl.c in upstart whose 
 * Author: Scott James Remnant <scott@netsplit.com>
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License version 2, as
 * published by the Free Software Foundation.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License along
 * with this program; if not, write to the Free Software Foundation, Inc.,
 * 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
 */

#ifdef HAVE_CONFIG_H
# include <config.h>
#endif /* HAVE_CONFIG_H */


#include <dbus/dbus.h>

#include <sys/types.h>

#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <unistd.h>
#include <fnmatch.h>

#include <nih/macros.h>
#include <nih/alloc.h>
#include <nih/string.h>
#include <nih/main.h>
#include <nih/option.h>
#include <nih/command.h>
#include <nih/logging.h>
#include <nih/error.h>
#include <nih/hash.h>
#include <nih/tree.h>

#include <nih-dbus/dbus_error.h>
#include <nih-dbus/dbus_proxy.h>
#include <nih-dbus/errors.h>
#include <nih-dbus/dbus_connection.h>

#include "dbus/mountall.h"

#include "com.ubuntu.Mountall.Server.h"


/* Prototypes for option and command functions */
int stop_timer_action		(NihCommand *command, char * const *args);
int restart_timer_action	(NihCommand *command, char * const *args);
int change_mount_device_action  (NihCommand *command, char * const *args);

/**
 * dest_name:
 *
 * Name on the D-Bus system bus that the message should be sent to when
 * system is TRUE.
 **/
char *dest_name = NULL;

/**
 * mountall_open:
 * @parent: parent object for new proxy.
 *
 * Opens a connection to the init daemon and returns a proxy to the manager
 * object.  If @dest_name is not NULL, a connection is instead opened to
 * the system bus and the proxy linked to the well-known name given.
 *
 * Error messages are output to standard error.
 *
 * If @parent is not NULL, it should be a pointer to another object which
 * will be used as a parent for the returned proxy.  When all parents
 * of the returned proxy are freed, the returned proxy will also be
 * freed.
 *
 * Returns: newly allocated D-Bus proxy or NULL on error.
 **/
NihDBusProxy *
mountall_open (const void *parent)
{
	DBusError       dbus_error;
	DBusConnection *connection;
	NihDBusProxy *  mountall;
	int uid = getuid ();

	dbus_error_init (&dbus_error);
	if (uid) {
		nih_error ("Need to be root to execute this command ");
		return NULL;
	}	
		
	connection = dbus_connection_open (DBUS_ADDRESS_MNTALL, &dbus_error);
	if (! connection) {
		nih_error ("%s: %s", _("Unable to connect to mountall"),
			   dbus_error.message);
		dbus_error_free (&dbus_error);
		return NULL;
	}
	dbus_error_free (&dbus_error);

	mountall = nih_dbus_proxy_new (parent, connection,
				      dest_name,
				      DBUS_PATH_MNTALL,
				      NULL, NULL);
	if (! mountall) {
		NihError *err;

		err = nih_error_get ();
		nih_error ("%s", err->message);
		nih_free (err);

		dbus_connection_unref (connection);
		return NULL;
	}
	nih_debug("Proxy for Mountall0_1.Server created!");
	mountall->auto_start = FALSE;

	/* Drop initial reference now the proxy holds one */
	dbus_connection_unref (connection);
	return mountall;
}

/**
 * stop_timer_action:
 * @command: NihCommand invoked,
 * @args: command-line arguments.
 *
 * This function is called for the "stop-timer" command.
 *
 * Returns: 1 on error and 0 on success.
 **/
int
stop_timer_action (NihCommand *  command,
		   char * const *args)
{
	nih_local NihDBusProxy *mountall = NULL;
	NihError *              err;

	nih_assert (command != NULL);
	nih_assert (args != NULL);

	if (!args[0]) {
		fprintf (stderr, _("%s: missing device name \n"), program_name);
		nih_main_suggest_help ();
		return 1;
	}
	mountall = mountall_open (NULL);
	if (! mountall)
		return 1;
	if (mountall_server_stop_timer_sync (NULL, mountall, args[0]))
		goto error;
	return 0;
error:
	err = nih_error_get ();
	nih_error ("%s", err->message);
	nih_free (err);
	return 1;
}

/**
 * restart_timer_action:
 * @command: NihCommand invoked,
 * @args: command-line arguments.
 *
 * This function is called for the "restart-timer" command.
 *
 * Returns: 1 on error and 0 on success.
 **/
int
restart_timer_action (NihCommand *  command,
		   char * const *args)
{
	nih_local NihDBusProxy *mountall = NULL;
	NihError *              err;

	nih_assert (command != NULL);
	nih_assert (args != NULL);

	if (!args[0]) {
		fprintf (stderr, _("%s: missing device name \n"), program_name);
		nih_main_suggest_help ();
		return 1;
	}
	mountall = mountall_open (NULL);
	if (! mountall)
		return 1;
	
	if (mountall_server_restart_timer_sync (NULL, mountall, args[0]))
		goto error;
	return 0;
error:
	err = nih_error_get ();
	nih_error ("%s", err->message);
	nih_free (err);
	return 1;
}

/**
 * change_mount_dev_action:
 * @command: NihCommand invoked,
 * @args: command-line arguments.
 *
 * This function is called for the "change-mount" command.
 *
 * Returns: 1 on error and 0 on success.
 **/
int
change_mount_dev_action (NihCommand *  command,
		   char * const *args)
{
	nih_local NihDBusProxy *mountall = NULL;
	NihError *              err;

	nih_assert (command != NULL);
	nih_assert (args != NULL);

	if (!args[0]) {
		fprintf (stderr, _("%s: missing device name \n"), program_name);
		nih_main_suggest_help ();
		return 1;
	}
	if (!args[1]) {
		fprintf (stderr, _("%s: missing mount point \n"), program_name);
		nih_main_suggest_help ();
		return 1;
	}
	mountall = mountall_open (NULL);
	if (! mountall)
		return 1;
	if (mountall_server_change_mount_device_sync (NULL, mountall, 
						args[0], args[1])) 
		goto error;
	return 0;
error:
	err = nih_error_get ();
	nih_error ("%s", err->message);
	nih_free (err);
	return 1;
}

/**
 * version_action:
 * @command: NihCommand invoked,
 * @args: command-line arguments.
 *
 * This function is called for the "version" command.
 *
 * Returns: command exit status.
 **/
int
version_action (NihCommand *  command,
		char * const *args)
{
	nih_local NihDBusProxy *mountall = NULL;
	nih_local char *        version = NULL;
	NihError *              err;

	nih_assert (command != NULL);
	nih_assert (args != NULL);

	mountall = mountall_open (NULL);
	if (! mountall)
		return 1;

	if (mountall_server_get_version_sync (NULL, mountall, &version) < 0)
		goto error;

	nih_message ("%s", version);

	return 0;

error:
	err = nih_error_get ();
	nih_error ("%s", err->message);
	nih_free (err);

	return 1;
}

#ifndef TEST
/**
 * options:
 *
 * Command-line options accepted for all arguments.
 **/
static NihOption options[] = {
	NIH_OPTION_LAST
};


NihOption stop_timer_options[] = {
	NIH_OPTION_LAST
};

NihOption restart_timer_options[] = {
	NIH_OPTION_LAST
};

NihOption change_mnt_dev_options[] = {
	NIH_OPTION_LAST
};

/**
 * version_options:
 *
 * Command-line options accepted for the version command.
 **/
NihOption version_options[] = {
	NIH_OPTION_LAST
};

/**
 * timer_group:
 *
 * Group of commands related to the timer
 **/
static NihCommandGroup timer_commands = { N_("Timer") };

/**
 * mounts_group:
 *
 * Group of commands related to mount devices and paths.
 **/
static NihCommandGroup mount_commands = { N_("Mounts") };


/**
 * commands:
 *
 * Commands accepts as the first non-option argument, or program name.
 **/
static NihCommand commands[] = {
	{ "stop-timer", N_("MOUNTPOINT"),
	  N_("Stop a timer associated with the specified device"),
	  N_("MOUNTPOINT is the mountpoint corresponding to a device whose "
	     "timer you want to stop"),
	  &timer_commands, stop_timer_options, stop_timer_action },

	{ "restart-timer", N_("MOUNTPOINT"),
	  N_("Restart a timer associated with the specified device"),
	  N_("MOUNTPOINT is the mountpoint corresponding to a device that "
	     "should become ready withing the default wait time or the time "
	     "specified explicitly as a command line argument to mountall") ,
	  &timer_commands, restart_timer_options, restart_timer_action },

	{ "change-mount", N_("DEVICE-NAME PATH"),
	  N_("Change the device to mount at a given mountpoint specified in "
	     "/etc/fstab" ),
	  N_("DEVICE-NAME is the name of the new device that you want to mount "
	     "PATH is the full path specified in /etc/fstab") ,
	  &mount_commands, change_mnt_dev_options, change_mount_dev_action },

	{ "version", NULL,
	  N_("Request the version of the mountall daemon."),
	  NULL,
	  NULL, version_options, version_action },

	NIH_COMMAND_LAST
};

int
main (int   argc,
      char *argv[])
{
	int ret;

	nih_main_init (argv[0]);

	ret = nih_command_parser (NULL, argc, argv, options, commands);
	if (ret < 0)
		exit (1);

	dbus_shutdown ();

	return ret;
}
#endif
